File: examples/timetabling/constraints/hard/RoomCapacity.ts
/**
 * HC3: Room capacity must accommodate all participants
 */

import type { Constraint } from 'timetable-sa';
import type { TimetableState } from '../../types/index.js';

export class RoomCapacity implements Constraint<TimetableState> {
  name = 'Room Capacity';
  type = 'hard' as const;

  evaluate(state: TimetableState): number {
    const { schedule, rooms } = state;
    const roomMap = new Map(rooms.map(r => [r.Code, r]));

    for (const entry of schedule) {
      const room = roomMap.get(entry.room);

      if (!room) {
        return 0; // Room not found - violation
      }

      if (room.Capacity < entry.participants) {
        return 0; // Room too small - violation
      }
    }

    return 1; // All rooms have sufficient capacity
  }

  describe(state: TimetableState): string | undefined {
    const { schedule, rooms } = state;
    const roomMap = new Map(rooms.map(r => [r.Code, r]));

    for (const entry of schedule) {
      const room = roomMap.get(entry.room);

      if (!room) {
        return `Room ${entry.room} not found for class ${entry.classId}`;
      }

      if (room.Capacity < entry.participants) {
        return `Room ${entry.room} capacity (${room.Capacity}) is less than participants (${entry.participants}) for class ${entry.classId}`;
      }
    }

    return undefined;
  }
}

File: python/src/timetable_sa/examples/timetabling/constraints/hard/room_capacity.py
"""Hard constraint: Room Capacity.

Ensures class participants do not exceed room capacity.
"""

from typing import Literal, List

from timetable_sa.core.interfaces import Constraint
from timetable_sa.examples.timetabling.domain_types.state import TimetableState, ScheduleEntry
from timetable_sa.examples.timetabling.domain_types.domain import Room


class RoomCapacity(Constraint):
    """Hard constraint: Class participants must not exceed room capacity.
    
    This ensures that classes are assigned to rooms that can accommodate
    all registered students.
    
    Examples of violations:
        - Class with 35 students in Room with capacity 30
        - Large lecture in small seminar room
    """
    
    name: str = "Room Capacity"
    type: Literal["hard", "soft"] = "hard"

    def evaluate(self, state: TimetableState) -> float:
        """Evaluate the constraint.
        
        Returns:
            1.0 if all classes fit in their rooms, lower score otherwise
        """
        violations = self._find_violations(state)
        
        if not violations:
            return 1.0
        
        return 1.0 / (1 + len(violations))

    def get_violations(self, state: TimetableState) -> List[str]:
        """Get detailed list of capacity violations.
        
        Returns:
            List of violation descriptions
        """
        violations = self._find_violations(state)
        descriptions = []
        
        for entry, room in violations:
            overflow = entry.participants - room.capacity
            descriptions.append(
                f"Class {entry.class_id} ({entry.kelas}) has {entry.participants} students "
                f"but room {room.code} capacity is {room.capacity} (overflow: {overflow})"
            )
        
        return descriptions

    def _find_violations(self, state: TimetableState) -> List[tuple[ScheduleEntry, Room]]:
        """Find all capacity violations.
        
        Returns:
            List of (entry, room) tuples where capacity is exceeded
        """
        violations = []
        
        # Build room lookup
        room_by_code = {r.code: r for r in state.rooms}
        
        for entry in state.schedule:
            room = room_by_code.get(entry.room)
            if room and entry.participants > room.capacity:
                violations.append((entry, room))
        
        return violations
